#### Two-Tailed One-Sample Z-Test

**1 - What is it?**

A two-tailed one-sample Z-test checks whether the sample mean is significantly different (either higher or lower) from the population mean, when the population standard deviation is known and sample size is larger (n>=30).

**In other words:** It checks the smaple mean deviates from the population mean in **either direction** not just greater or smaller, but any significant difference.

**Example:** Is my sample group's average score significantly different from the country's average either higher or lower is the difference due to random variation?

**2 - Real-Life Example: Two-Tailed One-Sample Z-Test**

`Scenario:`
Your product Manager claims: "The new onboarding flow changed the average session time. We don't know if it went up or down, but it definitely changed."

So we test:
- Is the enew *different* from 5 minutes?
- Not specifically higher or lower -> *two-tailed*.

**3 - Hypotheses**

- H0 (Null): u = 5 (No change)
- H1 (Alternate): u != 5 (There is some change: either increase or decrease)

This is *two-tailed* because deviation on *both sides* matters.

**4 - Formula & Parameters**

Z = (Sample Mean - Population Mean)/Standard Error
Standard Error (SE) = Population std/Sample size

**Parameters:**
- Population mean (u): 5 min
- Population std: 1 min
- Sample mean (x bar): 5.5 min
- Sample size: 50
- Significance level(alpha): 0.05
- Confidence Interval: 95%

Here, **population std is known and sample size>=30**, So we can do z-test, specifically *two-tailed one-sample z-test.*

*Difference from one-tailed:* 
- **Two tailed splits alpha into half:** alpha/2 = 0.025 on each side
- Two-tailed critical z-values are **+-1.96**

**5 - Python Implementation (Step-by-Step)**

In [10]:
import numpy as np
from scipy import stats

# ------------------------------------------------------------
# Step 1: Inputs
# ------------------------------------------------------------

population_mean = 5     #Expected average session time
population_std  = 1     #Population standard deviation
sample_mean     = 5.5   #Observed mean after change
sample_size     = 50    #Number of users
alpha           = 0.05  #5% significance level (two-tailed)

# ------------------------------------------------------------
# Step 2: Standard Error
# ------------------------------------------------------------

standard_error = population_std/np.sqrt(sample_size)
print(f"Standard Error: {standard_error:.4f}")

# ------------------------------------------------------------
# Step 3: Z-stat
# ------------------------------------------------------------

z_stat = (sample_mean - population_mean)/standard_error
print(f"Z-Statistics: {z_stat:.4f}")

# ------------------------------------------------------------
# Step 4: P-Value (Two-Tailed) 
# ------------------------------------------------------------
# For two-tailed: multiply by 2 because we check both sides

p_value = 2 * (1 - stats.norm.cdf(abs(z_stat)))
print(f"P-Value: {p_value:.4f}")

# ------------------------------------------------------------
# Step 5: Critical Values (Two-Tailed)
# ------------------------------------------------------------

critical_value_right = stats.norm.ppf(1-alpha/2)
critical_value_left = stats.norm.ppf(alpha/2)
print(f"Critical Values: Left={critical_value_left:.4f}, Right={critical_value_right:.4f}")

# ------------------------------------------------------------
# Step 6: Decision Based on P-Value
# ------------------------------------------------------------

print("\n----------- Decision Based on P-Value ---------------")
if p_value < alpha:
    print("Reject H0: The onboarding flow significantly changed session time.")
else:
    print("Fail to Reject H0: No strong evidence of change in session time.")

print("\n----------- Decision Based on Z vs Critical Values ------------")
if z_stat < critical_value_left or z_stat > critical_value_right:
    print("Reject H0: Statistically significant differences detected.")
else:
    print("Fail to Reject H0: No statistically meaningful differences.")

Standard Error: 0.1414
Z-Statistics: 3.5355
P-Value: 0.0004
Critical Values: Left=-1.9600, Right=1.9600

----------- Decision Based on P-Value ---------------
Reject H0: The onboarding flow significantly changed session time.

----------- Decision Based on Z vs Critical Values ------------
Reject H0: Statistically significant differences detected.


**6 - Business Interpretation**

- **Z = 1.90** -> The sample mean is 1.9 standard error below the population mean.
- **p = 0.029** -> There's only about a 2.9% probability of observing such a low average if the true mean were still 70.

**Conclusion:**
This suggests that decrease in score is **statistically significant**, meaning it's unlikely due to random variation.
The new syllabus **likely had a negative impact** on students performance.

**6 - Interpretation (Business Level)**

- **Z = 3.53** -> The sample mean is 3.5 standard error away from population mean.
- **p = 0.0004** -> There's only about a 0.04% probability of observing such a session time changes due to random noise

**Conclusion:** *Since 3.53>1.96 -> falls in rejection zone, The new onboarding flow changed session time significantly.*